File: VulkanGameEngineCore/CBuffer.c
#include "CBuffer.h"
#include "CVulkanRenderer.h"

 VkResult Buffer_UpdateBufferMemory(VkDevice device, VkDeviceMemory bufferMemory, void* dataToCopy, VkDeviceSize bufferSize)
{
    if (dataToCopy == NULL || bufferSize == 0)
    {
        RENDERER_ERROR("Buffer data and size cannot be NULL");
        return VK_ERROR_MEMORY_MAP_FAILED;
    }

    void* mappedData;
    VkResult result = vkMapMemory(device, bufferMemory, 0, bufferSize, 0, &mappedData);
    if (result != VK_SUCCESS)
    {
        RENDERER_ERROR("Failed to map buffer memory");
        return result;
    }

    memcpy(mappedData, dataToCopy, (size_t)bufferSize);
    vkUnmapMemory(device, bufferMemory);
    return VK_SUCCESS;
    {
    }
 }

 void Buffer_CopyBufferMemory(VkDevice device, VkCommandPool commandPool, VkDeviceMemory srcBuffer, VkDeviceMemory* dstBuffer, VkDeviceSize bufferSize)
 {
     VkBufferCopy copyRegion =
     {
         .srcOffset = 0,
         .dstOffset = 0,
         .size = bufferSize
     };

     VkCommandBuffer commandBuffer = Renderer_BeginSingleUseCommandBuffer(device, commandPool);
     vkCmdCopyBuffer(commandBuffer, srcBuffer, dstBuffer, 1, &copyRegion);
     Renderer_EndCommandBuffer(&commandBuffer);
 }

VkResult Buffer_AllocateMemory(VkDevice device, VkPhysicalDevice physicalDevice, VkBuffer* bufferData, VkDeviceMemory* bufferMemory, VkMemoryPropertyFlags properties)
{
    if (bufferData == NULL || bufferMemory == NULL)
    {
        RENDERER_ERROR("Buffer data or buffer memory is NULL");
        return VK_ERROR_MEMORY_MAP_FAILED;
    }

    VkMemoryRequirements memRequirements;
    vkGetBufferMemoryRequirements(device, *bufferData, &memRequirements);

    VkMemoryAllocateFlagsInfoKHR ExtendedAllocFlagsInfo =
    {
        .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_FLAGS_INFO_KHR
    };
    VkMemoryAllocateInfo allocInfo =
    {
        .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
        .allocationSize = memRequirements.size,
        .memoryTypeIndex = Renderer_GetMemoryType(physicalDevice, memRequirements.memoryTypeBits, properties),
        .pNext = &ExtendedAllocFlagsInfo,
    };

    return vkAllocateMemory(device, &allocInfo, NULL, bufferMemory);
}

void* Buffer_MapBufferMemory(VkDevice device, VkDeviceMemory bufferMemory, VkDeviceSize bufferSize, bool* isMapped)
{
    if (*isMapped)
    {
        RENDERER_ERROR("Buffer already mapped!");
        return NULL;
    }

    void* mappedData;
    VkResult result = vkMapMemory(device, bufferMemory, 0, bufferSize, 0, &mappedData);
    if (result != VK_SUCCESS)
    {
        RENDERER_ERROR("Failed to map buffer memory");
        return NULL;
    }
    *isMapped = true;
    return mappedData;
}

VkResult Buffer_UnmapBufferMemory(VkDevice device, VkDeviceMemory bufferMemory, bool* isMapped)
{
    if (*isMapped)
    {
        vkUnmapMemory(device, bufferMemory);
        *isMapped = false;
    }
    return VK_SUCCESS;
}

VkResult Buffer_CreateBuffer(VkDevice device, VkPhysicalDevice physicalDevice, VkBuffer* buffer, VkDeviceMemory* bufferMemory, void* bufferData, VkDeviceSize bufferSize, VkMemoryPropertyFlags properties, VkBufferUsageFlags usage)
{
    VkBufferCreateInfo bufferCreateInfo =
    {
        .sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
        .size = bufferSize,
        .usage = usage,
        .sharingMode = VK_SHARING_MODE_EXCLUSIVE
    };

    VkResult result = vkCreateBuffer(device, &bufferCreateInfo, NULL, buffer);
    result = Buffer_AllocateMemory(device, physicalDevice, buffer, bufferMemory, properties);
    result = vkBindBufferMemory(device, *buffer, *bufferMemory, 0);
    result = Buffer_UpdateBufferMemory(device, *bufferMemory, bufferData, bufferSize);

    return VK_SUCCESS;
}

VkResult Buffer_CreateStagingBuffer(VkDevice device, VkPhysicalDevice physicalDevice, VkCommandPool commandPool, VkQueue graphicsQueue, VkBuffer* stagingBuffer, VkBuffer* buffer, VkDeviceMemory* stagingBufferMemory, VkDeviceMemory* bufferMemory, void* bufferData, VkDeviceSize bufferSize, VkBufferUsageFlags bufferUsage, VkMemoryPropertyFlags properties)
{
    if (stagingBuffer == NULL || buffer == NULL || stagingBufferMemory == NULL || bufferMemory == NULL)
    {
        RENDERER_ERROR("One or more buffer pointers are NULL");
        return VK_ERROR_INITIALIZATION_FAILED;
    }

    VkResult result = Buffer_CreateBuffer(device, physicalDevice, stagingBuffer, stagingBufferMemory, bufferData, bufferSize, properties, bufferUsage);
    if (result != VK_SUCCESS)
    {
        return result;
    }

    result = Buffer_CreateBuffer(device, physicalDevice, buffer, bufferMemory, bufferData, bufferSize, properties, bufferUsage);
    if (result != VK_SUCCESS)
    {
        RENDERER_ERROR("Failed to create buffer");
        vkDestroyBuffer(device, *stagingBuffer, NULL);
        Renderer_FreeDeviceMemory(device, stagingBufferMemory);
        return result;
    }

    return Buffer_CopyBuffer(device, commandPool, graphicsQueue, stagingBuffer, buffer, bufferSize);
}

VkResult Buffer_CopyBuffer(VkDevice device, VkCommandPool commandPool, VkQueue graphicsQueue, VkBuffer* srcBuffer, VkBuffer* dstBuffer, VkDeviceSize size)
{
    if (srcBuffer == NULL || dstBuffer == NULL)
    {
        RENDERER_ERROR("Source or Destination Buffer is NULL");
        return VK_ERROR_UNKNOWN;
    }

    VkBufferCopy copyRegion =
    {
        .srcOffset = 0,
        .dstOffset = 0,
        .size = size
    };

    VkCommandBuffer commandBuffer = Renderer_BeginSingleUseCommandBuffer(device, commandPool);

    vkCmdCopyBuffer(commandBuffer, *srcBuffer, *dstBuffer, 1, &copyRegion);

    return Renderer_EndSingleUseCommandBuffer(device, commandPool, graphicsQueue, commandBuffer);
}

VkResult Buffer_UpdateBufferSize(VkDevice device, VkPhysicalDevice physicalDevice, VkBuffer buffer, VkDeviceMemory* bufferMemory, void* bufferData, VkDeviceSize* oldBufferSize, VkDeviceSize newBufferSize, VkBufferUsageFlags bufferUsageFlags, VkMemoryPropertyFlags propertyFlags)
{
    if (newBufferSize < *oldBufferSize)
    {
        RENDERER_ERROR("New buffer size can't be less than the old buffer size.");
        return VK_ERROR_MEMORY_MAP_FAILED;
    }

    *oldBufferSize = newBufferSize;

    VkBufferCreateInfo bufferCreateInfo =
    {
        .sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
        .size = newBufferSize,
        .usage = bufferUsageFlags,
        .sharingMode = VK_SHARING_MODE_EXCLUSIVE
    };

    VkBuffer newBuffer;
    VkDeviceMemory newBufferMemory;

    VkResult result = vkCreateBuffer(device, &bufferCreateInfo, NULL, &newBuffer);
    if (result != VK_SUCCESS)
    {
        RENDERER_ERROR("Failed to create new buffer");
        return result;
    }

    result = Buffer_AllocateMemory(device, physicalDevice, &newBuffer, &newBufferMemory, propertyFlags);
    if (result != VK_SUCCESS)
    {
        RENDERER_ERROR("Failed to allocate memory for new buffer");
        vkDestroyBuffer(device, newBuffer, NULL);
        return result;
    }

    result = vkBindBufferMemory(device, newBuffer, newBufferMemory, 0);
    if (result != VK_SUCCESS)
    {
        RENDERER_ERROR("Failed to bind memory to the new buffer");
        Renderer_FreeDeviceMemory(device, &newBufferMemory);
        vkDestroyBuffer(device, newBuffer, NULL);
        return result;
    }

    if (bufferData != NULL)
    {
        result = Buffer_UpdateBufferMemory(device, newBufferMemory, bufferData, newBufferSize);
        if (result != VK_SUCCESS)
        {
            RENDERER_ERROR("Failed to update memory with buffer data");
            vkDestroyBuffer(device, newBuffer, NULL);
            Renderer_FreeDeviceMemory(device, &newBufferMemory);
            return result;
        }
    }

    vkDestroyBuffer(device, buffer, NULL);
    Renderer_FreeDeviceMemory(device, bufferMemory);

    buffer = newBuffer;
    *bufferMemory = newBufferMemory;

    return VK_SUCCESS;
}

void Buffer_UpdateBufferData(VkDevice device, VkDeviceMemory* bufferMemory, void* dataToCopy, VkDeviceSize bufferSize, bool usingStagingBuffer)
{
    if (usingStagingBuffer == true)
    {
        RENDERER_ERROR("This buffer uses staging");
        return VK_ERROR_INITIALIZATION_FAILED;
    }

    Buffer_UpdateBufferMemory(device, *bufferMemory, dataToCopy, bufferSize);
}

void Buffer_UpdateStagingBufferData(VkDevice device, VkCommandPool commandPool, VkDeviceMemory* stagingBufferMemory, VkDeviceMemory* bufferMemory, void* dataToCopy, VkDeviceSize bufferSize, bool usingStagingBuffer)
{

    VkResult result = Buffer_UpdateBufferMemory(device, *stagingBufferMemory, dataToCopy, bufferSize);
    if (result != VK_SUCCESS)
    {
        return;
    }
    Buffer_CopyBufferMemory(device, commandPool, *stagingBufferMemory, bufferMemory, bufferSize);
}

VkResult Buffer_DestroyBuffer(VkDevice device, VkBuffer* buffer, VkBuffer* stagingBuffer, VkDeviceMemory* bufferMemory, VkDeviceMemory* stagingBufferMemory, void* bufferData, VkDeviceSize* bufferSize, VkBufferUsageFlags* bufferUsageFlags, VkMemoryPropertyFlags* propertyFlags)
{
    if (buffer == NULL && stagingBuffer == NULL)
    {
        RENDERER_ERROR("Both buffer and stagingBuffer are NULL");
        return VK_ERROR_INITIALIZATION_FAILED;
    }

    if (buffer && *buffer != VK_NULL_HANDLE)
    {
        vkDestroyBuffer(device, *buffer, NULL);
        *buffer = VK_NULL_HANDLE;
    }

    if (stagingBuffer && *stagingBuffer != VK_NULL_HANDLE)
    {
        vkDestroyBuffer(device, *stagingBuffer, NULL);
        *stagingBuffer = VK_NULL_HANDLE;
    }

    if (bufferMemory && *bufferMemory != VK_NULL_HANDLE)
    {
        Renderer_FreeDeviceMemory(device, bufferMemory);
        *bufferMemory = VK_NULL_HANDLE;
    }

    if (stagingBufferMemory && *stagingBufferMemory != VK_NULL_HANDLE)
    {
        Renderer_FreeDeviceMemory(device, stagingBufferMemory);
        *stagingBufferMemory = VK_NULL_HANDLE;
    }

    *bufferSize = 0;
    *bufferUsageFlags = 0;
    *propertyFlags = 0;
    bufferData = VK_NULL_HANDLE;

    return VK_SUCCESS;
}

File: VulkanGameEngineLevelEditor/Vulkan/CBuffer.cs
using Silk.NET.Vulkan;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using VulkanGameEngineLevelEditor.Vulkan;

namespace VulkanGameEngineLevelEditor.Vulkan
{
    public static unsafe class CBuffer
    {
        private static Vk vk = Vk.GetApi();
        public static VkResult AllocateMemory(VkBuffer buffer, ref VkDeviceMemory bufferMemory, VkMemoryPropertyFlagBits properties)
        {
            bufferMemory = new VkDeviceMemory();
            VkFunc.vkGetBufferMemoryRequirements(VulkanRenderer.device, buffer, out VkMemoryRequirements memRequirements);

            VkMemoryAllocateFlagsInfo extendedAllocFlagsInfo = new VkMemoryAllocateFlagsInfo
            {
                sType = VkStructureType.VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_FLAGS_INFO
            };

            VkMemoryAllocateInfo allocInfo = new VkMemoryAllocateInfo
            {
                sType = VkStructureType.VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
                allocationSize = memRequirements.size,
                memoryTypeIndex = VulkanRenderer.GetMemoryType(memRequirements.memoryTypeBits, properties),
                pNext = &extendedAllocFlagsInfo
            };


            VkFunc.vkAllocateMemory(VulkanRenderer.device, &allocInfo, null, out bufferMemory);

            return VkResult.VK_SUCCESS;
        }

        public static unsafe void* MapBufferMemory(VkDeviceMemory bufferMemory, ulong bufferSize, bool* isMapped)
        {
            if (*isMapped)
            {
                throw new InvalidOperationException("Buffer already mapped!");
            }

            void* mappedData;
            VkResult result = VkFunc.vkMapMemory(VulkanRenderer.device, bufferMemory, 0, bufferSize, 0, &mappedData);
            if (result != VkResult.VK_SUCCESS)
            {
                return null;
            }

            *isMapped = true;
            return mappedData;
        }

        public static Result CUnmapBufferMemory(VkDeviceMemory bufferMemory, bool* isMapped)
        {
            if (*isMapped)
            {
                VkFunc.vkUnmapMemory(VulkanRenderer.device, bufferMemory);
                *isMapped = false;
            }
            return Result.Success;
        }

        public static VkResult CreateBuffer(out VkBuffer buffer, out VkDeviceMemory bufferMemory, void* bufferData, ulong bufferSize, VkBufferUsageFlagBits bufferUsage, VkMemoryPropertyFlagBits properties)
        {
            if (bufferData == null || bufferSize == 0)
            {
                throw new InvalidOperationException("Buffer Data and Size can't be NULL");
            }

            bufferMemory = new VkDeviceMemory();

            VkBufferCreateInfo bufferInfo = new VkBufferCreateInfo
            {
                sType = VkStructureType.VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
                size = bufferSize,
                usage = bufferUsage,
                sharingMode = VkSharingMode.VK_SHARING_MODE_EXCLUSIVE
            };
            VkResult result = VkFunc.vkCreateBuffer(VulkanRenderer.device, &bufferInfo, null, out VkBuffer bufferPtr);
            buffer = bufferPtr;

            result = AllocateMemory(buffer, ref bufferMemory, properties);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            result = VkFunc.vkBindBufferMemory(VulkanRenderer.device, buffer, bufferMemory, 0);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            void* mappedData;
            result = VkFunc.vkMapMemory(VulkanRenderer.device, bufferMemory, 0, bufferSize, 0, &mappedData);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            int intCount = (int)bufferSize / 4;
            int[] mappedDataArray = new int[intCount];
            Marshal.Copy((IntPtr)bufferData, mappedDataArray, 0, intCount);
            VkFunc.vkUnmapMemory(VulkanRenderer.device, bufferMemory);

            return VkResult.VK_SUCCESS;
        }

        public static VkResult CreateStagingBuffer(out VkBuffer stagingBuffer, out VkDeviceMemory stagingBufferMemory, IntPtr bufferData, ulong bufferSize, VkBufferUsageFlagBits bufferUsage, VkMemoryPropertyFlagBits properties)
        {
            stagingBufferMemory = new VkDeviceMemory();
            var imageInfo = new VkBufferCreateInfo()
            {
                sType = VkStructureType.VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
                size = bufferSize,
                usage = VkBufferUsageFlagBits.VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
                sharingMode = VkSharingMode.VK_SHARING_MODE_EXCLUSIVE
            };
            VkFunc.vkCreateBuffer(VulkanRenderer.device, &imageInfo, null, out stagingBuffer);

            VkFunc.vkGetBufferMemoryRequirements(VulkanRenderer.device, stagingBuffer, out VkMemoryRequirements memRequirements);

            AllocateMemory(stagingBuffer, ref stagingBufferMemory, properties);
            return VkFunc.vkBindBufferMemory(VulkanRenderer.device, stagingBuffer, stagingBufferMemory, 0);
        }

        public static VkResult CopyBuffer(VkBuffer srcBuffer, VkBuffer dstBuffer, ulong size)
        {
            VkBufferCopy copyRegion = new VkBufferCopy
            {
                srcOffset = 0,
                dstOffset = 0,
                size = size
            };

            VkCommandBuffer commandBuffer = VulkanRenderer.BeginSingleUseCommandBuffer();
            VkFunc.vkCmdCopyBuffer(commandBuffer, srcBuffer, dstBuffer, 1, &copyRegion);
            return VulkanRenderer.EndSingleUseCommandBuffer(commandBuffer);
        }

        public static Result CopyStagingBuffer(VkCommandBuffer commandBuffer, VkBuffer srcBuffer, VkBuffer dstBuffer, ulong size)
        {
            VkBufferCopy copyRegion = new VkBufferCopy
            {
                srcOffset = 0,
                dstOffset = 0,
                size = size
            };

            VkFunc.vkCmdCopyBuffer(commandBuffer, srcBuffer, dstBuffer, 1, &copyRegion);
            return Result.Success;
        }

        public static VkResult UpdateBufferSize(VkBuffer buffer, VkDeviceMemory bufferMemory, void* bufferData, ref ulong oldBufferSize, ulong newBufferSize, VkBufferUsageFlagBits bufferUsageFlags, VkMemoryPropertyFlagBits propertyFlags)
        {
            if (newBufferSize < oldBufferSize)
            {
                throw new InvalidOperationException("New buffer size can't be less than the old buffer size.");
            }

            oldBufferSize = newBufferSize;

            VkBufferCreateInfo bufferCreateInfo = new VkBufferCreateInfo
            {
                sType = VkStructureType.VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
               size = newBufferSize,
                usage = bufferUsageFlags,
                sharingMode = VkSharingMode.VK_SHARING_MODE_EXCLUSIVE
            };

            VkResult result = VkFunc.vkCreateBuffer(VulkanRenderer.device, &bufferCreateInfo, null, out VkBuffer tempbuffer);
            buffer = tempbuffer;
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            result = AllocateMemory(buffer, ref bufferMemory, propertyFlags);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            result = VkFunc.vkBindBufferMemory(VulkanRenderer.device, buffer, bufferMemory, 0);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            return VkFunc.vkMapMemory(VulkanRenderer.device, bufferMemory, 0, newBufferSize, 0, &bufferData);
        }

        public static VkResult UpdateBufferMemory(VkDeviceMemory bufferMemory, void* dataToCopy, ulong bufferSize)
        {
            if (dataToCopy == null || bufferSize == 0)
            {
                throw new InvalidOperationException("Buffer Data and Size can't be NULL");
            }

            void* mappedData;
            VkResult result = VkFunc.vkMapMemory(VulkanRenderer.device, bufferMemory, 0, bufferSize, 0, &mappedData);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            int intCount = (int)bufferSize / 4;
            int[] mappedDataArray = new int[intCount];
            Marshal.Copy((IntPtr)dataToCopy, mappedDataArray, 0, intCount);
            VkFunc.vkUnmapMemory(VulkanRenderer.device, bufferMemory);
            return VkResult.VK_SUCCESS;
        }


        public static VkResult BuffUpdateStagingBufferMemory(VkDeviceMemory bufferMemory, void* dataToCopy, ulong bufferSize)
        {
            if (dataToCopy == null || bufferSize == 0)
            {
                throw new InvalidOperationException("Buffer Data and Size can't be NULL");
            }

            void* mappedData;
            VkResult result = VkFunc.vkMapMemory(VulkanRenderer.device, bufferMemory, 0, bufferSize, 0, &mappedData);
            if (result != VkResult.VK_SUCCESS)
            {
                return result;
            }

            int intCount = (int)bufferSize / 4;
            int[] mappedDataArray = new int[intCount];
            Marshal.Copy((IntPtr)dataToCopy, mappedDataArray, 0, intCount);
            VkFunc.vkUnmapMemory(VulkanRenderer.device, bufferMemory);
            return VkResult.VK_SUCCESS;
        }
    }
}
